File: 3기/디자인패턴/최현주/SourceCode/CompositePattern/Directory.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

namespace CompositePattern.ex2
{
    /// <summary>
    /// Composite 역할
    /// Entry(Component)를 저장, Component를 관리하기위한 메소드를 가지고있다.
    /// </summary>
    public class Directory : Entry
    {
        private string name;
        private List<Entry> entryList = new List<Entry>();

        public Directory(string name)
        {
            this.name = name;
        }

        public override string GetName()
        {
            return name;
        }

        //단일 객체들의 크기를 더하여 사이즈를 계산.
        public override int GetSize()
        {
            int totalSize = 0;

            for (int i = 0; i < entryList.Count; i++)
            {
                totalSize += entryList[i].GetSize();
            }

            return totalSize;
        }

        public override void Add(Entry entry)
        {
            entryList.Add(entry);
        }

        public override void PrintList(string prefix)
        {
            Debug.Log(prefix + "/" + this.name);

            for (int i = 0; i < entryList.Count; i++)
            {
                Entry entry = entryList[i];
                entry.PrintList(prefix + "/" + name);
            }
        }
    }
}

File: 3기2/디자인패턴2/최혁/workspace/Composite/src/Directory.java
import java.util.ArrayList;
import java.util.Iterator;

public class Directory extends Entry {
	private String name;
	private ArrayList<Entry> directory = new ArrayList<Entry>(); // 안에 담길 Entry List

	public Directory(String name) {
		this.name = name;
	}

	@Override
	public String getName() {
		return name;
	}

	@Override
	public int getSize() {
		int size = 0;
		Iterator<Entry> it = directory.iterator();
		while (it.hasNext()) {
			Entry entry = it.next();
			size += entry.getSize();
		}
		return size;
	}

	public Entry add(Entry entry) {
		directory.add(entry);
		return this;
	}

	@Override
	protected void printList(String prefix) {
		System.out.println(prefix + "/" + this); // this == this.toString() { getName() + " (" + getSize() + ")"; }
		Iterator<Entry> it = directory.iterator();
		while (it.hasNext()) {
			Entry entry = (Entry) it.next();
			entry.printList(prefix + "/" + name);
		}
	}

}
